lib/questionlib.php

```diff
@@ -2594,3 +2076,726 @@ function get_next_version(int $questionbankentryid): ?int {
+    }
+
+    /**
+     * Creates a stamp that uniquely identifies this version of the question
+     *
+     * In future we want this to use a hash of the question data to guarantee that
+     * identical versions have the same version stamp.
+     *
+     * @param object $question
+     * @return string A unique version stamp
+     * @deprecated since Moodle 4.0
+     */
+    function question_hash($question) {
+        debugging('Function question_hash() has been deprecated without replacement.', DEBUG_DEVELOPER);
+        return make_unique_id_code();
+    }
+
+    /**
+     * Private method, only for the use of add_indented_names().
+     *
+     * Recursively adds an indentedname field to each category, starting with the category
+     * with id $id, and dealing with that category and all its children, and
+     * return a new array, with those categories in the right order.
+     *
+     * @param array $categories an array of categories which has had childids
+     *          fields added by flatten_category_tree(). Passed by reference for
+     *          performance only. It is not modfied.
+     * @param int $id the category to start the indenting process from.
+     * @param int $depth the indent depth. Used in recursive calls.
+     * @param int $nochildrenof
+     * @return array a new array of categories, in the right order for the tree.
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function flatten_category_tree(&$categories, $id, $depth = 0, $nochildrenof = -1) {
+        debugging('Function flatten_category_tree() has been deprecated and moved to qbank_managecategories plugin,
+    Please use qbank_managecategories\helper::flatten_category_tree() instead.', DEBUG_DEVELOPER);
+
+        // Indent the name of this category.
+        $newcategories = array();
+        $newcategories[$id] = $categories[$id];
+        $newcategories[$id]->indentedname = str_repeat('&nbsp;&nbsp;&nbsp;', $depth) .
+            $categories[$id]->name;
+
+        // Recursively indent the children.
+        foreach ($categories[$id]->childids as $childid) {
+            if ($childid != $nochildrenof) {
+                $newcategories = $newcategories + flatten_category_tree(
+                        $categories, $childid, $depth + 1, $nochildrenof);
+            }
+        }
+
+        // Remove the childids array that were temporarily added.
+        unset($newcategories[$id]->childids);
+
+        return $newcategories;
+    }
+
+    /**
+     * Format categories into an indented list reflecting the tree structure.
+     *
+     * @param array $categories An array of category objects, for example from the.
+     * @param int $nochildrenof
+     * @return array The formatted list of categories.
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function add_indented_names($categories, $nochildrenof = -1) {
```

Comment:
remove the deprecated code and point to the new function

Reply:
done, changed to public in managecategories.